control-plane/agents/src/bin/core/tests/volume/resize.rs

```diff
@@ -179,3 +193,191 @@ async fn resize_on_no_capacity_pool() {
+    let cluster = ClusterBuilder::builder()
+        .with_rest(false)
+        .with_agents(vec!["core"])
+        .with_io_engines(2)
+        .with_pool(0, "malloc:///p1?size_mb=200")
+        .with_pool(1, "malloc:///p1?size_mb=200")
+        .with_cache_period("1s")
+        .with_reconcile_period(Duration::from_secs(1), Duration::from_secs(1))
+        .with_options(|o| o.with_isolated_io_engine(true))
+        .build()
+        .await
+        .unwrap();
+
+    let vol_cli = cluster.grpc_client().volume();
+
+    // resize exceeding the capacity limit
+    grpc_resize_volume_with_limit(
+        &vol_cli,
+        Some(EXPANDED_SIZE - CAPACITY_LIMIT_DIFF),
+        Some(ReplyErrorKind::CapacityLimitExceeded {}),
+    )
+    .await;
+
+    // resize within the capacity limit
+    grpc_resize_volume_with_limit(&vol_cli, Some(EXPANDED_SIZE + CAPACITY_LIMIT_DIFF), None).await;
+}
+
+async fn grpc_resize_volume_with_limit(
+    volume_client: &dyn VolumeOperations,
+    capacity: Option<u64>,
+    expected_error: Option<ReplyErrorKind>,
+) {
+    let volume = volume_client
+        .create(
+            &CreateVolume {
+                uuid: "de3cf927-80c2-47a8-adf0-95c486bdd7b7".try_into().unwrap(),
+                size: SIZE,
+                replicas: 2,
+                thin: false,
+                ..Default::default()
+            },
+            None,
+        )
+        .await
+        .unwrap();
+
+    let result = volume_client
+        .resize(
+            &ResizeVolume {
+                uuid: volume.uuid().clone(),
+                requested_size: EXPANDED_SIZE,
+                cluster_capacity_limit: capacity,
+            },
+            None,
+        )
+        .await;
+
+    match result {
+        Ok(resized_volume) => {
+            assert!(resized_volume.spec().uuid == volume.spec().uuid);
+            assert!(resized_volume.spec().size == EXPANDED_SIZE);
+            volume_client
+                .destroy(
+                    &DestroyVolume {
+                        uuid: resized_volume.uuid().try_into().unwrap(),
+                    },
+                    None,
+                )
+                .await
+                .unwrap();
+            assert!(expected_error.is_none());
+        }
+        Err(e) => {
+            assert_eq!(expected_error, Some(e.kind)); // wrong error
+                                                      // Volume not needed anymore.
+            volume_client
+                .destroy(
+                    &DestroyVolume {
+                        uuid: volume.uuid().try_into().unwrap(),
+                    },
+                    None,
+                )
+                .await
+                .unwrap();
+        }
+    }
+}
+
+// Create lots of volume on the pool and then resize a volume.
+// We'll create single replica volumes and two 800 MiB pools are available, so
+// let's create about 60 volumes of 20MiB each so we have some space left for
+// expansion of one of the volume.
```

Comment:
Why not just 1 volume of 1.2GiB ?

Reply:
The idea behind this test is to ensure we don't hang up cpu by holding locks. Hence more volumes' spec locks, plus the top level lock on all specs.

Reply:
Sorry I don't get it, where are we holding locks?

Reply:
> Sorry I don't get it, where are we holding locks?

I'm referring to the lock during capacity limit checks. Below part:

```
                let specs = self.write();       <------------------- this
                let mut total: u64 = specs
                    .volumes
                    .values()
                    .filter_map(|v| {
                        let vol_locked = v.lock();      <------------------- and this
```

Reply:
I still don't see how this test is testing that? Seems any resize would exercise that code anyway?